File: common/src/main/java/cn/rubintry/common/utils/http/HttpExecutors.java
package cn.rubintry.common.utils.http;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

/**
 * @author logcat
 */
public class HttpExecutors {

    private final Executor mDiskIO;
    private final Executor mNetworkIO;
    private final Executor mMainThread;

    private static volatile HttpExecutors instance;

    public static HttpExecutors getInstance(){
        if(instance == null){
            synchronized (HttpExecutors.class){
                if(instance == null){
                    instance = new HttpExecutors();
                }
            }
        }
        return instance;
    }

    public HttpExecutors(Executor mDiskIO, Executor mNetworkIO, Executor mMainThread) {
        this.mDiskIO = mDiskIO;
        this.mNetworkIO = mNetworkIO;
        this.mMainThread = mMainThread;
    }

    public Executor diskIO() {
        return mDiskIO;
    }

    public Executor networkIO() {
        return mNetworkIO;
    }

    public Executor mainThread() {
        return mMainThread;
    }

    public HttpExecutors(){
        this(Executors.newSingleThreadExecutor() , Executors.newFixedThreadPool(3) , new MainThreadExecutors());
    }



    private static class MainThreadExecutors  implements Executor {

        private Handler handler = new Handler(Looper.getMainLooper());
        @Override
        public void execute(Runnable runnable) {
            handler.post(runnable);
        }
    }

}

File: common/src/main/java/cn/rubintry/common/utils/http/HttpExecutors.kt
package cn.rubintry.common.utils.http

import android.os.Handler
import android.os.Looper
import java.util.concurrent.Executor
import java.util.concurrent.Executors

/**
 * @author logcat
 */
class HttpExecutors @JvmOverloads constructor(
    private val mDiskIO: Executor = Executors.newSingleThreadExecutor(),
    private val mNetworkIO: Executor = Executors.newFixedThreadPool(
        3
    ),
    private val mMainThread: Executor = MainThreadExecutors()
) {
    fun diskIO(): Executor {
        return mDiskIO
    }

    fun networkIO(): Executor {
        return mNetworkIO
    }

    fun mainThread(): Executor {
        return mMainThread
    }

    private class MainThreadExecutors : Executor {
        private val handler = Handler(Looper.getMainLooper())
        override fun execute(runnable: Runnable) {
            handler.post(runnable)
        }
    }

    companion object {
        @Volatile
        var instance: HttpExecutors? = null
            get() {
                if (field == null) {
                    synchronized(HttpExecutors::class.java) {
                        if (field == null) {
                            field = HttpExecutors()
                        }
                    }
                }
                return field
            }
            private set
    }

}